text/0000-const-ub.md

```diff
@@ -0,0 +1,135 @@
+- Feature Name: `const_ub`
+- Start Date: 2020-10-10
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Define how UB during const evaluation is treated:
+some kinds of UB must be detected, the remaining UB conditions are ignored and evaluation continues in a well-defined way.
+However, CTFE queries causing UB are not subject to stability guarantees and thus may fail to build in the future (e.g. when more UB is being detected).
+
+# Motivation
+[motivation]: #motivation
+
+So far, nothing is specified about what happens when `unsafe` code leads to UB during CTFE.
+This is a major blocker for stabilizing `unsafe` operations in const-contexts.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+There are some values that Rust needs to compute at compile-time.
+This includes the initial value of a `const`/`static`, and array lengths (and more general, const generics).
+Computing these initial values is called compile-time function evaluation (CTFE).
+CTFE in Rust is very powerful and permits running almost arbitrary Rust code.
+This begs the question, what happens when there is `unsafe` code and it causes [Undefined Behavior (UB)][UB]?
+
+The answer depends on the kind of UB: some kinds of UB are guaranteed to be detected,
+while other kinds of UB might either be detected, or else evaluation will continue as if the violated UB condition did not exist (i.e., as if this operation was actually defined).
+This can change from compiler version to compiler version: CTFE code that causes UB could build fine with one compiler and fail to build with another.
+
+This RFC does not alter the general policy that unsound code is not subject to strict stability guarantees.
+In other words, unsafe code may not rely on all future versions of Rust to implement this RFC.
+The RFC only helps *consumers* of unsafe code to be sure that right now, all UB during CTFE will be detected or non-consequential.
```

Comment:
what is "non-consequential" supposed to mean here? Does it mean "it actually *is* defined behavior at the moment, and the definition of that behavior is 'the obvious thing' [whatever that means] for the given CTFE expression"?

Reply:
Actually I think the reference section does answer this question, namely "All of this UB has in common that there is an "obvious" way to continue evaluation". I'm not 100% sure I agree that the meaning of dereferencing an unaligned pointer is **obvious** though.

Reply:
> I'm not 100% sure I agree that the meaning of dereferencing an unaligned pointer is obvious though.

Maybe I can word this better; suggestions are welcome. From the perspective of a formal spec or a Miri-like interpreter, alignment constraints are of the form
```rust
if something_is_wrong { raise_ub; }
/* go on */
```
So there is an "obvious" way to continue evaluation even when something is wrong: just skip the `if` and go on without raising UB.

This is in stark contrast to e.g. OOB memory accesses, if we do not raise UB, we'll simply ICE later since we'd index a `Vec<u8>` (representing this memory allocation) beyond its bounds.

@pnkfelix suggestions for better wording are welcome. :)

Reply:
@RalfJung let me put it to you this way: does the part labeled `/* go on */` explicitly mask out the bits causing mis-alignment, so that the resulting address *does* obey the alignment constraints?

Because the heart of my complaint is that I don't know whether the "obvious" interpretation of an unaligned access is to implicitly mask out the bits to get the desired (e.g. word) alignment, or if it is to access the two neighboring words and then combine their bytes to construct the "word" that was denoted by the misaligned address.

Reply:
> Because the heart of my complaint is that I don't know whether the "obvious" interpretation of an unaligned access is to implicitly mask out the bits to get the desired (e.g. word) alignment, or if it is to access the two neighboring words and then combine their bytes to construct the "word" that was denoted by the misaligned address.

Ah, interesting. The option of doing such masking never occurred to me.

In the interpreter it would be extra work to mask off the "overaligned" bits, so we do nothing like that -- we just read the given number of bytes starting at the given address, ignoring the fact that the alignment constraint is violated. No need to read neighboring things and combine them.

> If you tell me that your intent is that all non-signaled cases of UB will be handled in some similar specified manner, then I'll be happy.

Looks like we should do that then. ;) What would be the place for such documentation though?

Reply:
> Ah, interesting. The option of doing such masking never occurred to me.

(The most common place in my career when I have encountered misaligned pointers was when there was some form of type tag embedded in the value. Its pretty common in dynamic language runtimes. That's where the masking interpretation originates for me.)